craft_parts/callbacks.py

```diff
@@ -0,0 +1,140 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Register and execute callback functions."""
+
+import logging
+from collections import namedtuple
+from typing import Callable, List, Union
+
+from craft_parts import errors
+from craft_parts.infos import ProjectInfo, StepInfo
+from craft_parts.parts import Part
+from craft_parts.steps import Step
+
+CallbackHook = namedtuple("CallbackHook", ["function", "step_list"])
+
+ExecutionCallback = Callable[[ProjectInfo, List[Part]], None]
+StepCallback = Callable[[StepInfo], bool]
+Callback = Union[ExecutionCallback, StepCallback]
+
+_PROLOGUE_HOOKS: List[CallbackHook] = []
+_EPILOGUE_HOOKS: List[CallbackHook] = []
+_PRE_STEP_HOOKS: List[CallbackHook] = []
+_POST_STEP_HOOKS: List[CallbackHook] = []
+
+logger = logging.getLogger(__name__)
+
+
+def register_prologue(func: ExecutionCallback) -> None:
```

Comment:
All of these hooks should likely be combined into a context object and available as properties rather than being global.  Then the idea of "clear()" is no longer required.

Reply:
My initial idea was to have a callback manager instantiated by the lifecycle manager, and this callback manager would hold the lists. A similar approach could be used for plugins. However, the typical use case for both callbacks and plugins, if they're used at all, is to register them once in the caller application and never unregister -- so static registration seemed to simplify the API because callbacks and plugins could then be registered at any point of the code (e.g. at start of execution), without having to wait for the lifecycle manager to be created. I'll think about it again.